src/backend/gporca/libgpopt/src/base/CUtils.cpp

```diff
@@ -926,34 +932,49 @@ CUtils::CollectConsumersAndProducers(CMemoryPool *mp, CExpression *pexpr,
 		if (!pexprChild->Pop()->FScalar())
 		{
 			CollectConsumersAndProducers(mp, pexprChild, cteConsumers,
-										 cteProducerSet);
+										 cteProducerSet, false);
 		}
 	}
 }
 
 BOOL
 CUtils::hasUnpairedCTEConsumer(CMemoryPool *mp, CExpression *pexpr)
 {
-	BOOL hasUnpairedConsumer = false;
-
-	ULongPtrArray *cteConsumers = GPOS_NEW(mp) ULongPtrArray(mp);
 	UlongCteIdHashSet *cteProducerSet = GPOS_NEW(mp) UlongCteIdHashSet(mp);
+	UlongCteIdHashSet *cteConsumerSet = GPOS_NEW(mp) UlongCteIdHashSet(mp);
+
+	CollectConsumersAndProducers(mp, pexpr, cteConsumerSet, cteProducerSet,
+								 true);
 
-	CollectConsumersAndProducers(mp, pexpr, cteConsumers, cteProducerSet);
+	// hazzard slice must contains all producers, else here is unpaired
+	// cte consumers/producers
+	if (cteConsumerSet->Size() != cteProducerSet->Size())
+	{
+		return true;
```

Comment:
You forgot to free memory for cteConsumerSet and cteProducerSet!
No wonder there used to be a return of a local variable!

Reply:
My bad, i've forgot about that